File: leetcode/javascript/56-merge.js
/**
 * 以数组 intervals 表示若干个区间的集合，
 * 其中单个区间为 intervals[i] = [starti, endi] 。
 * 请你合并所有重叠的区间，并返回一个不重叠的区间数组，
 * 该数组需恰好覆盖输入中的所有区间 。
 */

/**
 * @description 合并区间
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    for (let i = 0; i < intervals.length; ++i) {
        if (!intervals[i]) {
            continue;
        }

        for (let j = i + 1; j < intervals.length; ++j) {
            if (!intervals[j]) {
                continue;
            }

            if (!intervals[i]) {
                break;
            }

            var min = Math.min(intervals[i][0], intervals[j][0]);
            var max = Math.max(intervals[i][1], intervals[j][1]);

            var sumLen = intervals[i][1] - intervals[i][0] + intervals[j][1] - intervals[j][0];

            if (max - min <= sumLen) {
                intervals[i] = [min, max];
                intervals[j] = null;
                --i;
            }
        }
    }

    return intervals.filter(val => !!val);
};

/**
 * 测试用例：
 * console.log(merge([[2,3],[4,5],[6,7],[8,9],[1,10]]))
 */

/**
 * 本题核心： 无
 *
 * 反思：普通处理时间复杂度为O(n ^ 2)，可采用先排序后合并的方式
 */

File: leetcode/python/56-merge.py
from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        ans=[intervals[0]]
        for i in intervals[1:]:
            if ans[-1][1]>=i[0]:
                #有交集算并集,因为已经升序排列了,所以左端点必比当前区间左端点小,不需要维护判断。
                ans[-1]=[ans[-1][0],max(ans[-1][1],i[1])]
            else:
                #无交集直接加
                ans.append(i)
        return ans
        
print(Solution().merge([[2,3],[4,5],[6,7],[8,9],[1,10]]))
print(Solution().merge([[1,3],[2,6],[8,10],[15,18]]))
print(Solution().merge([[5,5],[1,3],[3,5],[4,6],[1,1],[3,3],[5,6],[3,3],[2,4],[0,0]]))

